src/main/java/graphql/schema/impl/SchemaUtil.java

```diff
@@ -0,0 +1,99 @@
+     */
+
+    public static void visitPartiallySchema(final GraphQLSchema partiallyBuiltSchema, GraphQLTypeVisitor... visitors) {
+        List<GraphQLSchemaElement> roots = new ArrayList<>();
+        roots.add(partiallyBuiltSchema.getQueryType());
+
+        if (partiallyBuiltSchema.isSupportingMutations()) {
+            roots.add(partiallyBuiltSchema.getMutationType());
+        }
+
+        if (partiallyBuiltSchema.isSupportingSubscriptions()) {
+            roots.add(partiallyBuiltSchema.getSubscriptionType());
+        }
+
+        if (partiallyBuiltSchema.getAdditionalTypes() != null) {
+            roots.addAll(partiallyBuiltSchema.getAdditionalTypes());
+        }
+
+        if (partiallyBuiltSchema.getDirectives() != null) {
+            roots.addAll(partiallyBuiltSchema.getDirectives());
+        }
+
+        roots.add(partiallyBuiltSchema.getIntrospectionSchemaType());
+
+        GraphQLTypeVisitor visitor = new MultiReadOnlyGraphQLTypeVisitor(Arrays.asList(visitors));
+        SchemaTraverser traverser;
+        traverser = new SchemaTraverser(schemaElement -> schemaElement.getChildrenWithTypeReferences().getChildrenAsList());
+        traverser.depthFirst(visitor, roots);
+    }
+
+    public static ImmutableMap<String, List<GraphQLObjectType>> groupInterfaceImplementationsByName(List<GraphQLNamedType> allTypesAsList) {
+        Map<String, List<GraphQLObjectType>> result = new LinkedHashMap<>();
+        for (GraphQLType type : allTypesAsList) {
+            if (type instanceof GraphQLObjectType) {
+                List<GraphQLNamedOutputType> interfaces = ((GraphQLObjectType) type).getInterfaces();
+                for (GraphQLNamedOutputType interfaceType : interfaces) {
+                    List<GraphQLObjectType> myGroup = result.computeIfAbsent(interfaceType.getName(), k -> new ArrayList<>());
+                    myGroup.add((GraphQLObjectType) type);
+                }
+            }
+        }
+        return ImmutableMap.copyOf(new TreeMap<>(result));
+    }
+
+    public Map<String, List<GraphQLImplementingType>> groupImplementationsForInterfacesAndObjects(GraphQLSchema schema) {
+        Map<String, List<GraphQLImplementingType>> result = new LinkedHashMap<>();
+        for (GraphQLType type : schema.getAllTypesAsList()) {
+            if (type instanceof GraphQLImplementingType) {
+                List<GraphQLNamedOutputType> interfaces = ((GraphQLImplementingType) type).getInterfaces();
+                for (GraphQLNamedOutputType interfaceType : interfaces) {
+                    List<GraphQLImplementingType> myGroup = result.computeIfAbsent(interfaceType.getName(), k -> new ArrayList<>());
+                    myGroup.add((GraphQLImplementingType) type);
+                }
+            }
+        }
+        return ImmutableMap.copyOf(new TreeMap<>(result));
+    }
+
+    public static void replaceTypeReferences(GraphQLSchema schema) {
+        final Map<String, GraphQLNamedType> typeMap = schema.getTypeMap();
```

Comment:
I am not sure about this code: it looks suspicious that we grab all types to iterate. I think we can do better and maybe this is not even completely correct. But we can leave it for another PR.